### PR TITLE
fix: `MergeDeep` for typescript 5.4

### LINKED ISSUE #784: Canary: `MergeDeep` failures in `typescript@next` not detected by canary tests
I'm seeing a lot of these style of errors in the canary tests in my own modules:

```
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(144,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownArrayOrTuple'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(144,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownRecord'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(144,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Source>>>[number]' and 'UnknownArrayOrTuple'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(170,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownArrayOrTuple'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(170,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownRecord'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(170,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Source>>>[number]' and 'UnknownArrayOrTuple'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(196,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownArrayOrTuple'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(196,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownRecord'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(196,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Source>>>[number]' and 'UnknownArrayOrTuple'.
```

I didn't receive them in `typescript@5.4.0-dev.20231119` then started receiving them in `typescript@5.4.0-dev.20231121`.

To avoid a regression once `typescript@5.4.0` gets launched, we should track down the cause of this and file an issue with `typescript` if needed.

We should also track down why the canary tests in `type-fest` itself isn't catching it.

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/784">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/784/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by skarab42**
Closes #806 
Fixes #784

### CODE REVIEW

**Comment by skarab42** [comment]
Canary run : [sindresorhus/type-fest/actions/runs/7699763916](https://github.com/sindresorhus/type-fest/actions/runs/7699763916)

**Review by sindresorhus — APPROVED** [review]

File: source/merge-deep.d.ts [hunk 1]
```diff
@@ -76,9 +79,12 @@ type Rest5 = PickRestType<string[]>; // => string[]
 ```
 */
 type PickRestType<Type extends UnknownArrayOrTuple> = number extends Type['length']
-	? ArrayTail<Type> extends [] ? Type : PickRestType<ArrayTail<Type>>
+	? PickRestType___<ArrayTail<Type>, Type>
 	: [];
 
+// Utility to avoid computing ArrayTail twice.
+type OmitRestType___<Tail extends UnknownArrayOrTuple, Type extends UnknownArrayOrTuple, Result extends UnknownArrayOrTuple = []> = Tail extends [] ? Result : OmitRestType<Tail, [...Result, FirstArrayElement<Type>]>;
```
Comment (voxpelli) [inline source/merge-deep.d.ts:86]: Can we maybe break this up over a few more lines – its hard for the eyes to follow the flow I think

**Review by voxpelli — APPROVED** [review]
It seems to work, which is great, a neater solution than mine 👍

File: source/merge-deep.d.ts (hunk 1 above)
Comment (skarab42) [reply source/merge-deep.d.ts]: Done ;)

**Comment by sindresorhus** [comment]
Thanks for the fix @skarab42. And thanks for the review @voxpelli 👍

**Pull request merged**